File: samples/ios-cmp-app/iosApp/Engine Wrapper/AuthenticatedResourceProvider.swift
import BaseCmpApp
import WebViewEngine

class AuthenticatedResourceProvider: ResourceProvider {
    private let authManager: AuthManager

    init(authManager: AuthManager) {
        self.authManager = authManager
    }

    func shouldHandle(request: URLRequest) -> Bool {
        true
    }

    func performRequest(_ request: URLRequest) async throws -> (Data, URLResponse) {
        if let token = try await authManager.getAccessToken() {
            try await URLSession.shared.data(for: request.authorizedRequest(with: token))
        } else {
            // neither token returned nor error thrown - let's try to make request without token
            try await URLSession.shared.data(for: request)
        }
    }
}

extension URLRequest {
    fileprivate func authorizedRequest(with token: String) -> URLRequest {
        var mutableRequest = self
        mutableRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        return mutableRequest
    }
}

File: samples/base-cmp-app/src/iosMain/kotlin/com/pega/constellation/sdk/kmp/samples/basecmpapp/ios/AuthenticatedResourceProvider.kt
package com.pega.constellation.sdk.kmp.samples.basecmpapp.ios

import com.pega.constellation.sdk.kmp.engine.webview.ios.ResourceProvider
import com.pega.constellation.sdk.kmp.samples.basecmpapp.auth.AuthManager
import platform.Foundation.NSHTTPURLResponse
import platform.Foundation.NSMutableURLRequest
import platform.Foundation.NSURLRequest
import platform.Foundation.NSURLResponse
import platform.Foundation.NSURLSession
import platform.Foundation.dataTaskWithRequest
import platform.Foundation.setValue
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class AuthenticatedResourceProvider(val authManager: AuthManager) : ResourceProvider {
    override fun shouldHandle(request: NSURLRequest) = true

    override suspend fun performRequest(request: NSURLRequest) =
        performNetworkRequest(
            authManager.getAccessToken()
                ?.let { request.authorizedRequest(it) }
                ?: request
        )

    private suspend fun performNetworkRequest(request: NSURLRequest) =
        suspendCoroutine { continuation ->
            NSURLSession.sharedSession.dataTaskWithRequest(request) { data, response, error ->
                if (response.isUnauthorized()) {
                    authManager.onTokenExpired()
                }

                when {
                    error != null -> continuation.resumeWithException(Exception(error.localizedDescription))
                    data != null && response != null -> continuation.resume(data to response)
                    else -> continuation.resumeWithException(Exception("Unknown error"))
                }
            }.resume()
        }

    private fun NSURLResponse?.isUnauthorized(): Boolean {
        return (this as? NSHTTPURLResponse)?.statusCode?.toInt() == 401
    }

    fun NSURLRequest.authorizedRequest(token: String) = (this.mutableCopy() as NSMutableURLRequest)
        .apply { setValue("Bearer $token", forHTTPHeaderField = "Authorization") }
}